make the remove icon of the file preview configurable and default to RemoveCircle

PennyJeans opened the pull request:
Closes #8755

fzaninotto commented:
Thanks for your PR. What's the use case for this?

fzaninotto commented:
OK, can you please PR against `next` rather than `master` as it's a new feature, and update the documentation to mention the new prop?

fzaninotto commented:
Bump, would you mind revisiting this PR?

PennyJeans commented:
Thank you, I changed the base to `next` and updated the docs. Also I refactored a little, so that the default icon is set similarly to the way it is already done in [BulkDeleteWithUndoButton ](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/button/BulkDeleteWithUndoButton.tsx).

slax57 commented on docs/FileInput.md:
```diff
@@ -62,6 +62,7 @@ Files are accepted or rejected based on the `accept`, `multiple`, `minSize` and
 | `multiple`             | Optional | `boolean`           | `false`    | Whether the inputs can accept multiple files.                       |
 | `options`              | Optional | `Object`            | `{}`       | Additional options passed to react-dropzone's `useDropzone()` hook. |
 | `placeholder`          | Optional | `ReactNode`         | -          | Invite displayed in the drop zone                                   |
+| `removeIcon`           | Optional | `ReactNode`         | -          | The clickable icon for removing images                              |
```
```suggestion
| `removeIcon`           | Optional | `ReactNode`         | [MUI's RemoveCircle icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) | The clickable icon for removing files                              |
```

slax57 commented on docs/FileInput.md:
````diff
@@ -152,6 +153,10 @@ If that's not enough, you can pass a `placeholder` prop to overwrite it. The val
 </FileInput>
 ```
 
+## `removeIcon`
+
+Optionally overwrite the [default icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) for removing files.
````
Could you add a small code example showing how one can use `removeIcon` to change the default icon to something else? (take inspiration from the other sections of the doc :wink:)

PennyJeans replied:
Done, and I tried to keep it consistent with the other sections.

slax57 commented on docs/ImageInput.md:
```diff
@@ -53,13 +53,14 @@ Files are accepted or rejected based on the `accept`, `multiple`, `minSize` and
 | `multiple`             | Optional | `boolean`           | `false`    | Whether the inputs can accept multiple files.                       |
 | `options`              | Optional | `Object`            | `{}`       | Additional options passed to react-dropzone's `useDropzone()` hook. |
 | `placeholder`          | Optional | `ReactNode`         | -          | Invite displayed in the drop zone                                   |
+| `removeIcon`           | Optional | `ReactNode`         | -          | The clickable icon for removing images                              |
```
```suggestion
| `removeIcon`           | Optional | `ReactNode`         | [MUI's RemoveCircle icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) | The clickable icon for removing images                              |
```

PennyJeans replied:
done

slax57 commented on packages/ra-ui-materialui/src/input/FileInputPreview.tsx:
```diff
@@ -30,7 +38,7 @@ export const FileInputPreview = (props: FileInputPreviewProps) => {
                 size="small"
             >
-                <RemoveCircle className={FileInputPreviewClasses.removeIcon} />
+                {removeIcon}
```
I'm afraid you're no longer passing the `className` here, which is a BC. You should pass it to whatever React Node `removeIcon` is.

PennyJeans replied:
Thanks for pointing this out. There are different ways to fix this and I now pushed a version where the icon is handed in as a component and not a react element. I hope this is fine for you.

slax57 reviewed (changes requested):
Also, can you add a story to both `FileInput.stories.tsx` and `ImageInput.stories.tsx` to demonstrate this new prop? Thanks!

PennyJeans commented on docs/ImageInput.md:
```diff
@@ -149,6 +150,10 @@ If that's not enough, you can pass a `placeholder` prop to overwrite it. The val
 
+## `removeIcon`
+
+Optionally overwrite the [default icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) for removing images.
```
done

slax57 commented on docs/FileInput.md:
```diff
@@ -152,6 +153,16 @@ If that's not enough, you can pass a `placeholder` prop to overwrite it. The val
 
+## `removeIcon`
+
+Optionally overwrite the [default icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) for removing files:
```
```suggestion
Use the `removeIcon` prop to change the icon displayed as the remove button:
```

slax57 commented on docs/FileInput.md:
````diff
@@ -152,6 +153,16 @@ If that's not enough, you can pass a `placeholder` prop to overwrite it. The val
+Optionally overwrite the [default icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) for removing files:
+
+```jsx
+<ImageInput source="attachments" RemoveIcon={CustomSvgIcon}>
+    <ImageField source="src" title="title" />
+</ImageInput>
````
```suggestion
<FileInput source="attachments" removeIcon={CustomSvgIcon}>
    <FileField source="src" title="title" />
</FileInput>
```

slax57 commented on docs/FileInput.md:
```diff
@@ -53,16 +53,17 @@ Files are accepted or rejected based on the `accept`, `multiple`, `minSize` and
+| `multiple`             | Optional | `boolean`           | `false`   | Whether the inputs can accept multiple files.                       |
+| `options`              | Optional | `Object`            | `{}`      | Additional options passed to react-dropzone's `useDropzone()` hook. |
+| `placeholder`          | Optional | `ReactNode`         | -         | Invite displayed in the drop zone                                   |
+| `removeIcon`           | Optional | `ReactNode`         | [MUI's RemoveCircle icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) | The clickable icon for removing images                              |
```
```suggestion
| `removeIcon`           | Optional | `ReactNode`         | [MUI's RemoveCircle icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) | The clickable icon for removing files                              |
```

slax57 commented on docs/ImageInput.md:
```diff
@@ -149,6 +150,16 @@ If that's not enough, you can pass a `placeholder` prop to overwrite it. The val
 
+## `removeIcon`
+
+Optionally overwrite the [default icon](https://mui.com/material-ui/material-icons/?query=removeCir&selected=RemoveCircle) for removing images:
```
```suggestion
Use the `removeIcon` prop to change the icon displayed as the remove button:
```

slax57 commented on packages/ra-ui-materialui/src/input/FileInputPreview.tsx:
```diff
@@ -42,6 +50,7 @@ FileInputPreview.propTypes = {
     className: PropTypes.string,
     file: PropTypes.object,
     onRemove: PropTypes.func.isRequired,
+    removeIcon: PropTypes.element,
```
Careful with the prop's casing. I'd suggest using lower camel case everywhere (i.e. `removeIcon` instead of `RemoveIcon`), here and in `FileInput.tsx`

slax57 replied:
I agree that the post you linked does recommend to use upper-casing for this case, and this is probably a good practice indeed.

However, in the case of react-admin, there are already many places in the code where component props are lower-case, hence for the sake of code consistency I'd rather have the `removeIcon` prop lower-case too.

You can find examples with [the `Layout.tsx` component](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/layout/Layout.tsx).

[Code](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/layout/Layout.tsx#L22-L32):

```tsx
    const {
        appBar: AppBar = DefaultAppBar,
        children,
        className,
        dashboard,
        error: errorComponent,
        menu: Menu = DefaultMenu,
        sidebar: Sidebar = DefaultSidebar,
        title,
        ...rest
    } = props;
```

[Documented usage](https://marmelab.com/react-admin/Layout.html#usage):

```tsx
import { Layout } from 'react-admin';

import { MyAppBar } from './MyAppBar';

export const MyLayout = props => <Layout {...props} appBar={MyAppBar} />;
```

PennyJeans replied:
Ok, thanks for the clarification! I changed the prop to camel case everywhere.

slax57 commented on packages/ra-ui-materialui/src/input/ImageInput.stories.tsx:
```diff
@@ -83,6 +84,15 @@ export const Required = () => (
+export const CustomRemoveIcon = () => (
+    <Wrapper>
+        <ImageInput source="image" RemoveIcon={DeleteIcon}>
+            <FileField source="src" title="title" />
```
You should probably use `ImageField` here instead of `FileField`, just like the other stories of this file (and remove the unnecessary import).

slax57 reviewed (changes requested):
The stories are working fine :slightly_smiling_face:  (except the image preview which I pointed out).

And I'm happy with the React component vs. element solution :slightly_smiling_face:

slax57 reviewed (approved):
Nice, thanks!
